core/drivers/crypto/caam/hal/common/hal_ctrl.c

```diff
@@ -10,50 +10,82 @@
 #include <registers/ctrl_regs.h>
 #include <registers/version_regs.h>
 
-uint8_t caam_hal_ctrl_jrnum(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_era(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
 
-	val = io_caam_read32(baseaddr + CHANUM_MS);
+	/* Read the number of instance */
+	val = io_caam_read32(baseaddr + CCBVID);
 
-	return GET_CHANUM_MS_JRNUM(val);
+	return GET_CCBVID_CAAM_ERA(val);
 }
 
-uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_jrnum(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
+	uint8_t jrnum;
 
-	/* Read the number of instance */
-	val = io_caam_read32(baseaddr + CHANUM_LS);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
+		val = io_caam_read32(baseaddr + CHANUM_MS);
 
-	if (GET_CHANUM_LS_MDNUM(val)) {
-		/* Hashing is supported */
-		val = io_caam_read32(baseaddr + CHAVID_LS);
-		val &= BM_CHAVID_LS_MDVID;
-		if (val == CHAVID_LS_MDVID_LP256)
-			return TEE_MAIN_ALGO_SHA256;
+		jrnum = GET_CHANUM_MS_JRNUM(val);
+	} else {
+		val = io_caam_read32(baseaddr + JR_VERSION);
 
-		return TEE_MAIN_ALGO_SHA512;
+		jrnum = GET_JR_VERSION_JRNUM(val);
 	}
 
-	return UINT8_MAX;
+	return jrnum;
 }
 
-uint8_t caam_hal_ctrl_pknum(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
 
-	val = io_caam_read32(baseaddr + CHANUM_LS);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
+		/* Read the number of instance */
+		val = io_caam_read32(baseaddr + CHANUM_LS);
+
+		if (GET_CHANUM_LS_MDNUM(val)) {
+			/* Hashing is supported */
+			val = io_caam_read32(baseaddr + CHAVID_LS);
+			val &= BM_CHAVID_LS_MDVID;
+			if (val == CHAVID_LS_MDVID_LP256)
+				return TEE_MAIN_ALGO_SHA256;
+
+			return TEE_MAIN_ALGO_SHA512;
+		}
+	} else {
+		/* Read the number of instance */
+		val = io_caam_read32(baseaddr + MDHA_VERSION);
 
-	return GET_CHANUM_LS_PKNUM(val);
+		if (GET_MDHA_VERSION_MDNUM(val)) {
+			/* Hashing is supported */
+			val &= BM_MDHA_VERSION_MDVID;
+			if (val == MDHA_VERSION_MDVID_LP256)
+				return TEE_MAIN_ALGO_SHA256;
+
+			return TEE_MAIN_ALGO_SHA512;
+		}
+	}
+
+	return UINT8_MAX;
 }
 
-uint8_t caam_hal_ctrl_era(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_pknum(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
+	uint8_t pknum;
 
-	/* Read the number of instance */
-	val = io_caam_read32(baseaddr + CCBVID);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
+		val = io_caam_read32(baseaddr + CHANUM_LS);
 
-	return GET_CCBVID_CAAM_ERA(val);
+		pknum = GET_CHANUM_LS_PKNUM(val);
+	} else {
+		val = io_caam_read32(baseaddr + PKHA_VERSION);
+
```

Comment:
Delete empty line

Reply:
I had added the blank line for clarity but I will remove it